cpp/CMakeLists.txt

```diff
@@ -389,6 +389,9 @@ if(BUILD_CUML_CPP_LIBRARY)
 
   # single GPU components
   add_library(${CUML_CPP_TARGET} SHARED
+    src/specializations/prims/selection/knn.cu
```

Comment:
I love the execution of how you have proposed and organized this! 

The only things I can think off the top of my head are:

- Add somewhere a readme that lists what functions have been specialized (maybe a markdown in the C++) to make it very easy and obvious how to use/reference/add specializations, unless you think this would be redundant. I think it could be valuable if we get a significantly large number of specializations. 
- Add a section to the C++ developer guide with usage guides and also guidelines of when a new specialization is preferred over importing the prims/raft functions directly. 

What do you think?

Reply:
I think both of those are great ideas! I've written scripts to help automate generation of the specializations, so I'll add something to them to start generating the markdown list as well.